### PR TITLE
implemented directories

### PR DESCRIPTION
**Pull request by TheCataliasTNT2k**
## Description
- implemented directories ("is_directory" is used to indicated if a file is a directory)
- added "parent_dir_uuid" to link to the parrent directory of the file 
- added boolean changeable to File, to prevent "/" file from being moved or deleted
- device creation also creates "/"
- directories can no longer moved inside itself or its child directories
- not changeable files can no longer be changed

## Related Issue
#46 

## Motivation and Context
Now there are directories not only files

## How Has This Been Tested?
not yet tested, unit tests will follow

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.

### CODE REVIEW

**Review by Defelo — CHANGES_REQUESTED** [review]
In the `["file", "create"]` endpoint you should check if the parent directory is actually a directory (done). Currently it is possible to create files in other files. Also it should not be possible to create files outside of the root directory, so `parent_dir_uuid` may not be `None` (done)

**Comment by Defelo** [comment]
In the `["device", "delete"]` endpoint you should also delete all files on the device. (done)

File: app/resources/device.py [hunk 1]
```diff
@@ -274,6 +277,11 @@ def delete_user(data: dict, microservice: str) -> dict:
         for service in wrapper.session.query(Service).filter_by(device_uuid=device.uuid):
             wrapper.session.delete(service)
 
+        #delete all files in filesystem
+        files: list = wrapper.session.query(File).filter_by(device=device.uuid)
+        for file in files:
+            wrapper.session.delete(file)
+
```
Comment (Defelo) [inline app/resources/device.py]: You already deleted the files above

File: app/resources/file.py [hunk 2]
```diff
@@ -173,14 +248,23 @@ def create_file(data: dict, user: str) -> dict:
...
-        device=device.uuid, filename=filename
+        device=device.uuid, filename=filename, parent_dir_uuid=parent_dir_uuid
     ).scalar()
 
     if file_count > 0:
         return file_already_exists
 
-    file: File = File.create(device.uuid, filename, content)
+    parent_dir: File = wrapper.session.query(File).filter_by(device=device.uuid, parent_dir_uuid=parent_dir_uuid)
+    if not parent_dir:
```
Comment (Defelo) [inline app/resources/file.py]: You also need to check if the parent is actually a directory:
```suggestion
    if not (parent_dir and parent_dir.is_directory):
```

File: app/resources/file.py [hunk 3]
```diff
@@ -173,14 +248,24 @@ def create_file(data: dict, user: str) -> dict:
...
     file_count: int = wrapper.session.query(func.count(File.uuid)).filter_by(
-        device=device.uuid, filename=filename
+        device=device.uuid, filename=filename, parent_dir_uuid=parent_dir_uuid
     ).scalar()
 
     if file_count > 0:
         return file_already_exists
 
-    file: File = File.create(device.uuid, filename, content)
+    parent_dir: File = wrapper.session.query(File).filter_by(device=device.uuid, parent_dir_uuid=parent_dir_uuid,
```
Comment (Defelo) [inline app/resources/file.py]: wrong query
```suggestion
    parent_dir: File = wrapper.session.query(File).filter_by(device=device.uuid, uuid=parent_dir_uuid,
```

**Review by Defelo — APPROVED** [review]

File: app/schemes.py [hunk 4]
```diff
@@ -17,6 +17,18 @@ def make_error(*args: str, origin: str, sep: str = "") -> dict:
 
 file_already_exists: dict = make_error("file_already_exists", origin="user")
 
+directory_can_not_have_textcontent: dict = make_error("directory_can_not_have_textcontent", origin="user")
+
+parent_dir_does_not_exist: dict = make_error("parent directory for new file does not exist", origin="user")
```
Comment (Defelo) [inline app/schemes.py]: duplicate error message for `parent_directory_not_found` -> delete this

File: app/resources/file.py [hunk 5]
```diff
@@ -173,14 +248,25 @@ def create_file(data: dict, user: str) -> dict:
...
 
     if file_count > 0:
         return file_already_exists
 
-    file: File = File.create(device.uuid, filename, content)
+    parent_dir: File = wrapper.session.query(File).filter_by(
+        device=device.uuid, uuid=parent_dir_uuid, is_directory=True
+    ).first()
+    if not parent_dir:
+        return parent_dir_does_not_exist
```
Comment (Defelo) [inline app/resources/file.py]: use `parent_directory_not_found` instead

**Review by Defelo — CHANGES_REQUESTED** [review]

**Review by Defelo — APPROVED** [review]

**Review by Defelo — CHANGES_REQUESTED** [review]
Device names may only consist of alphanumeric characters plus `-` and `_`. Similarly file names my only consist of alphanumeric characters plus `-`, `_` and `.`. (done)

**Review by Defelo — APPROVED** [review]

**Review by MaxiHuHe04 — CHANGES_REQUESTED** [review]
https://github.com/cryptic-game/frontend/pull/82#discussion_r341569568
Please add a notification for when files get created, deleted and changed. It should contain the affected file(s).  The structure could be for example:
```json
{
    "notify-id": "file-update",
    "origin": "<create/delete/update [string]>",
    "device_uuid": "<UUID of the affected device [string, UUID]>",
    "data": {
        "created": ["<UUID of the created file [string, UUID]>"],
        "deleted": ["<UUID of the deleted file {string, UUID]>"],
        "changed": ["<UUID of the changed file [string, UUID]>"]
    }
}
```

**Review by MaxiHuHe04 — APPROVED** [review]

**Review by Defelo — CHANGES_REQUESTED** [review]
Remove the `/` directory. Instead for all files in the root directory `parent_dir_uuid` should be `None` (done)

**Review by Defelo — APPROVED** [review]

**Pull request merged**